Moar Py3

s0undt3ch opened the pull request.

s0undt3ch committed:
Only convert bytes to strings under Python 3

s0undt3ch committed:
Don't even try to concatenate non string types

s0undt3ch committed:
Minor lint fixes

s0undt3ch committed:
Respect the LazyLoader `mod_dict_class` attribute if defined.

s0undt3ch committed:
Enforce and preserve module loading order.

This is the only way we can have predictable results when overriding,
for example, grains.

salt-jenkins commented:
@s0undt3ch, thanks for your PR! By analyzing the history of the files in this pull request, we identified @jacksontj, @msteed and @bastiaanb to be potential reviewers.

s0undt3ch committed:
Don't even try to convert if not bytes, str or unicode

s0undt3ch committed:
Fix a weird import issue

```
ImportError: Failed to import test module: unit.modules.docker_test
Traceback (most recent call last):
  File "/usr/lib64/python2.7/unittest/loader.py", line 252, in _find_tests
    module = self._get_module_from_name(name)
  File "/usr/lib64/python2.7/unittest/loader.py", line 230, in _get_module_from_name
    __import__(name)
  File "/testing/tests/unit/modules/docker_test.py", line 39, in <module>
    class DockerTestCase(TestCase):
  File "/testing/tests/unit/modules/docker_test.py", line 103, in DockerTestCase
    @skipIf(_docker_py_version() < (1, 4, 0),
  File "/testing/tests/unit/modules/docker_test.py", line 34, in _docker_py_version
    return docker_mod.docker.version_info
AttributeError: 'module' object has no attribute 'version_info'
```

s0undt3ch commented:
I really don't know what's up with that docker test. @terminalmage, any insights?

s0undt3ch commented:
Unrelated failure. Merging.

ScoreUnder commented on salt/loader.py:
```diff
@@ -1177,10 +1177,10 @@ def refresh_file_mapping(self):
         self.file_mapping = salt.utils.odict.OrderedDict()
 
-        for mod_dir in self.module_dirs:
+        for mod_dir in sorted(self.module_dirs):
```
I know it's a little late now that this is merged, but this looks like a mistake to me. The `module_dirs` field is populated in order of priority, so the sorting will break this.

There are comments in the `_module_dirs` functions like this one:
```
        # Prepend to the list to match cli argument ordering
```
Which suggest that the author does not expect the list to be sorted afterwards.

s0undt3ch replied:
I created https://github.com/saltstack/salt/issues/40517 in order not to forget about this
